In [43]:
#18-35年龄筛选
#学历分组，四类，大学，中学，初中，小学，文盲
#用python复现逻辑回归


#做异质性？？？？？-年龄分组

In [2]:
import pandas as pd

In [3]:
df_price=pd.read_excel(r'./分省房价数据.xlsx')
df_price

省份  2021年房价
0        北京市    46941
1        天津市    16370
2        河北省     8330
3        山西省     6689
4     内蒙古自治区     6516
5        辽宁省     9049
6        吉林省     7050
7       黑龙江省     6242
8        上海市    40974
9        江苏省    13666
10       浙江省    20386
11       安徽省     7904
12       福建省    12653
13       江西省     7648
14       山东省     8743
15       河南省     6438
16       湖北省     9100
17       湖南省     6481
18       广东省    16453
19   广西壮族自治区     5992
20       海南省    17550
21       重庆市     9678
22       四川省     8304
23       贵州省     5623
24       云南省     7869
25     西藏自治区     8443
26       陕西省     9680
27       甘肃省     5985
28       青海省     7851
29   宁夏回族自治区     6916
30  新疆维吾尔自治区     5548

In [4]:
df=pd.read_excel(r'./CGSS2021.xlsx')

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8148 entries, 0 to 8147
Data columns (total 31 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Unnamed: 0    8148 non-null   int64 
 1   省份            8148 non-null   object
 2   家人数量（除自己）     8148 non-null   int64 
 3   出生年份          8148 non-null   int64 
 4   出生月份          8148 non-null   object
 5   出生日期          8148 non-null   object
 6   最高教育程度        8127 non-null   object
 7   最高教育程度的状态     7243 non-null   object
 8   获得最高学历时间      6397 non-null   object
 9   所上高中等级        3181 non-null   object
 10  所上大学等级        1720 non-null   object
 11  大学专业（最后所学）    1720 non-null   object
 12  个人2020年收入     8148 non-null   object
 13  所住房屋面积        8148 non-null   object
 14  是否幸福          5458 non-null   object
 15  希望有几个孩子       8148 non-null   object
 16  希望有几个儿子       8148 non-null   object
 17  希望有几个女儿       8148 non-null   object
 18  2020家庭年收入     8148 non-null   object
 19  有几个儿子（

In [6]:
# Perform a left join
df['省份'] = df['省份'].astype(str)
df = pd.merge(df, df_price, on='省份', how='left')
df

Unnamed: 0   省份  家人数量（除自己）  出生年份 出生月份 出生日期        最高教育程度 最高教育程度的状态  \
0              0  北京市          3  1985   11    1  大学本科（成人高等教育）        毕业   
1              1  北京市          3  1957    2   16          普通高中        毕业   
2              2  北京市          1  1961    9    7          普通高中        毕业   
3              3  北京市          1  1937   11    8           NaN        毕业   
4              4  北京市          2  1968    8   15  大学本科（正规高等教育）        毕业   
...          ...  ...        ...   ...  ...  ...           ...       ...   
8143        8143  四川省          2  1991   10    7  大学专科（正规高等教育）        毕业   
8144        8144  四川省          5  2002   10    7  大学本科（成人高等教育）       正在读   
8145        8145  四川省          4  1978    5   12            初中        毕业   
8146        8146  四川省          0  1993    9   10            初中   辍学和中途退学   
8147        8147  四川省          1  1946   12    8            小学        毕业   

     获得最高学历时间     所上高中等级  ... 有几个亲生女儿 未成年子女数   婚姻状况         伴侣同居情况 父亲出生年份  \
0        2019   县、市级重点中学  ...      没有      1  初婚有配偶           住在一起   1956   
1        1972    不分重点非重点  ...       1     没有  初婚有配偶           住在一起   1936   
2        1978    不分重点非重点  ...       1      1  初婚有配偶  不住在一起，但同在一个城市   1930   
3         不知道    其他（请注明）  ...       2     没有     丧偶            NaN   1908   
4        1991  省/直辖市重点中学  ...      没有     没有  初婚有配偶           住在一起   1937   
...       ...        ...  ...     ...    ...    ...            ...    ...   
8143     2012      非重点中学  ...     NaN    NaN     未婚            NaN   1967   
8144     2020   县、市级重点中学  ...     NaN    NaN     未婚            NaN   1979   
8145     1993        NaN  ...      没有     没有  初婚有配偶  不住在一起，但同在一个城市   1959   
8146     2005        NaN  ...     NaN    NaN     未婚            NaN   1970   
8147     1956        NaN  ...      没有     没有     同居           住在一起   1940   

            父亲教育程度 母亲出生年份    母亲教育程度 是否同意孩子在家中最重要 2021年房价  
0               初中   1956        初中      不同意也不反对   46941  
1               小学   1932  没有受过任何教育          不同意   46941  
2           私塾、扫盲班   1930    私塾、扫盲班           同意   46941  
3               其他   1908  没有受过任何教育           同意   46941  
4     大学本科（正规高等教育）   1938      职业高中           同意   46941  
...            ...    ...       ...          ...     ...  
8143            初中   1969        初中          不同意    8304  
8144          普通高中   1980        初中      不同意也不反对    8304  
8145            小学   1960        小学           同意    8304  
8146            小学   1971        小学           同意    8304  
8147      没有受过任何教育   1941  没有受过任何教育           同意    8304  

[8148 rows x 32 columns]

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8148 entries, 0 to 8147
Data columns (total 32 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Unnamed: 0    8148 non-null   int64 
 1   省份            8148 non-null   object
 2   家人数量（除自己）     8148 non-null   int64 
 3   出生年份          8148 non-null   int64 
 4   出生月份          8148 non-null   object
 5   出生日期          8148 non-null   object
 6   最高教育程度        8127 non-null   object
 7   最高教育程度的状态     7243 non-null   object
 8   获得最高学历时间      6397 non-null   object
 9   所上高中等级        3181 non-null   object
 10  所上大学等级        1720 non-null   object
 11  大学专业（最后所学）    1720 non-null   object
 12  个人2020年收入     8148 non-null   object
 13  所住房屋面积        8148 non-null   object
 14  是否幸福          5458 non-null   object
 15  希望有几个孩子       8148 non-null   object
 16  希望有几个儿子       8148 non-null   object
 17  希望有几个女儿       8148 non-null   object
 18  2020家庭年收入     8148 non-null   object
 19  有几个儿子（

In [8]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
from sklearn.preprocessing import MinMaxScaler

# Step 1: Load and clean the data
selected_columns = ['省份', '最高教育程度', '所住房屋面积', '个人2020年收入', '2021年房价','希望有几个孩子']
df_7001 = df[selected_columns].copy()

# Step 2: Handle non-numeric and missing values
df_7001['所住房屋面积'] = pd.to_numeric(df_7001['所住房屋面积'], errors='coerce')#
df_7001['个人2020年收入'] = pd.to_numeric(df_7001['个人2020年收入'], errors='coerce')
df_7001['希望有几个孩子'] = pd.to_numeric(df_7001['希望有几个孩子'], errors='coerce')
#df['2021年房价'] = pd.to_numeric(df['2021年房价'], errors='coerce')
#This function can handle non-numeric data, convert it to numeric types, and return NaN (Not a Number) if the conversion fails.

In [9]:
# Drop rows with missing or invalid data after conversion
df_7001.dropna(subset=['所住房屋面积', '个人2020年收入', '希望有几个孩子','2021年房价'], inplace=True)

# Step 3: Convert categorical variables to numeric，.astype('category').cat.codes，Education classification can be further processed
df_7001['省份'] = df_7001['省份'].astype('category').cat.codes
df_7001['最高教育程度'] = df_7001['最高教育程度'].astype('category').cat.codes

# Step 4: Normalize numerical columns (house area, personal income)
scaler = MinMaxScaler()
df_7001[['所住房屋面积', '个人2020年收入','2021年房价']] = scaler.fit_transform(df_7001[['所住房屋面积', '个人2020年收入','2021年房价']])

# Step 5: Convert target variable to binary (0 = fewer than 2 children, 1 = 2 or more children)
df_7001['高生育意愿'] = df_7001['希望有几个孩子'].apply(lambda x: 1 if x >= 2 else 0)

# Step 6: Define features (X) and target (y)
X = df_7001[['省份', '最高教育程度', '所住房屋面积', '个人2020年收入','2021年房价']]
y = df_7001['高生育意愿']

# Add a constant term to the model (intercept)
X = sm.add_constant(X)

# Step 7: Fit the logistic regression model using statsmodels
model = sm.Logit(y, X)
result = model.fit()

Optimization terminated successfully.
         Current function value: 0.493859
         Iterations 6


In [10]:
# Print the full model summary
print("\nFull Model Summary:")
print(result.summary())


Full Model Summary:
                           Logit Regression Results                           
Dep. Variable:                  高生育意愿   No. Observations:                 6398
Model:                          Logit   Df Residuals:                     6392
Method:                           MLE   Df Model:                            5
Date:                Wed, 13 Nov 2024   Pseudo R-squ.:                 0.02900
Time:                        16:07:24   Log-Likelihood:                -3159.7
converged:                       True   LL-Null:                       -3254.1
Covariance Type:            nonrobust   LLR p-value:                 7.167e-39
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.2987      0.105     12.372      0.000       1.093       1.504
省份            -0.0049      0.006     -0.798      0.425      -0.017       0.007
最高教育程度         0.0045      0.01

Detailed Interpretation:

	1.	Intercept (const = 1.2942):
	•	The positive intercept suggests a baseline tendency towards high fertility intention when all predictors are zero.
	•	The p-value (0.000) indicates that the intercept is highly significant.
	2.	Province (省份):
	•	Coefficient: -0.0049, Odds Ratio ≈ e^{-0.0049} \approx 0.995.
	•	Interpretation: The effect of province on fertility intention is very small and not statistically significant (p-value = 0.425). This suggests that regional differences, as captured by the province codes, do not have a significant effect on fertility intention in this model.
	3.	Education Level (最高教育程度):
	•	Coefficient: 0.0045, Odds Ratio ≈ e^{0.0045} \approx 1.005.
	•	Interpretation: The effect of education level on fertility intention is almost negligible and not statistically significant (p-value = 0.641). This implies that, in your dataset, education level does not significantly influence fertility intentions.
	4.	House Area (所住房屋面积):
	•	Coefficient: 4.8154, Odds Ratio ≈ e^{4.8154} \approx 123.6.
	•	Interpretation: A strong positive effect, indicating that larger house areas are strongly associated with higher fertility intention. The odds of having a high fertility intention increase dramatically as the house area increases.
	•	The p-value (0.000) indicates that this variable is highly significant.
	5.	Personal Income in 2020 (个人2020年收入):
	•	Coefficient: -1.1625, Odds Ratio ≈ e^{-1.1625} \approx 0.31.
	•	Interpretation: There is a negative association between personal income and fertility intention, but this effect is not statistically significant (p-value = 0.166). Higher income tends to reduce the likelihood of high fertility intention, though the evidence is not strong enough to be conclusive.
	6.	Housing Price in 2021 (2021年房价):
	•	Coefficient: -1.0659, Odds Ratio ≈ e^{-1.0659} \approx 0.34.
	•	Interpretation: There is a significant negative effect of housing prices on fertility intention. Higher housing prices are associated with a decrease in fertility intentions.
	•	The p-value (0.000) indicates that this variable is highly significant, suggesting that housing prices are a strong deterrent to having more children.

Key Insights:

	•	Significant Factors:
	•	House Area: Strong positive effect on fertility intention, indicating that families with larger houses are more likely to want more children.
	•	Housing Prices: Strong negative effect, suggesting that high housing prices may discourage people from having more children.
	•	Non-significant Factors:
	•	Province, Education Level, and Personal Income did not show significant effects on fertility intention in this model.

Recommendations:

	•	Policy Implication: If housing prices are indeed a strong deterrent, policies aimed at reducing housing costs or providing housing support could positively influence fertility intentions.
	•	Further Analysis: It might be worthwhile to explore interactions between income, education, and housing prices, as these factors might have a combined effect that isn’t captured in the current model.

In [47]:
df_7001

省份  最高教育程度    所住房屋面积  个人2020年收入   2021年房价  希望有几个孩子  高生育意愿
0      1       4  0.023627   0.006004  1.000000      2.0      1
1      1       8  0.038393   0.005004  1.000000      0.0      0
2      1       8  0.027171   0.003603  1.000000      1.0      0
3      1      -1  0.021855   0.006308  1.000000      3.0      1
4      1       5  0.075015   0.013009  1.000000      1.0      0
...   ..     ...       ...        ...       ...      ...    ...
8135   2       9  0.060839   0.001001  0.056622      2.0      1
8137   2       5  0.066745   0.000000  0.056622      2.0      1
8141   2       9  0.034259   0.007505  0.056622      0.0      0
8142   2       1  0.054932   0.003002  0.056622      2.0      1
8145   2       1  0.043119   0.004203  0.056622      2.0      1

[6398 rows x 7 columns]

In [49]:
X1 = df_7001[['最高教育程度', '所住房屋面积', '个人2020年收入','2021年房价']]
y1 = df_7001['高生育意愿']
# Add a constant term to the model (intercept)
X1 = sm.add_constant(X1)

# Step 7: Fit the logistic regression model using statsmodels
model = sm.Logit(y1, X1)
result1 = model.fit()
print("\nFull Model Summary:")
print(result1.summary())

Optimization terminated successfully.
         Current function value: 0.493909
         Iterations 6

Full Model Summary:
                           Logit Regression Results                           
Dep. Variable:                  高生育意愿   No. Observations:                 6398
Model:                          Logit   Df Residuals:                     6393
Method:                           MLE   Df Model:                            4
Date:                Wed, 13 Nov 2024   Pseudo R-squ.:                 0.02891
Time:                        17:01:32   Log-Likelihood:                -3160.0
converged:                       True   LL-Null:                       -3254.1
Covariance Type:            nonrobust   LLR p-value:                 1.337e-39
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.2427      0.078     15.961      0.000       1.090       1.395
最高教育程度  